packages/node_modules/@ciscospark/internal-plugin-ediscovery/src/ediscovery.js

```diff
@@ -0,0 +1,520 @@
+/*!
+ * Copyright (c) 2015-2017 Cisco Systems, Inc. See LICENSE file.
+ */
+
+import {waitForValue, SparkPlugin} from '@ciscospark/spark-core';
+import reportRequest from './report-request';
+import ReportGenerator from './report-generator';
+
+/**
+ * @class Ediscovery is used by compliance officers to run compliance reports
+ *
+ */
+const Ediscovery = SparkPlugin.extend({
+  namespace: 'Ediscovery',
+
+  /**
+   * Creates a compliance report with a specific set of search parameters
+   *
+   * @param {String} name - A label to identify the report
+   * @param {String} description - A textual summary of the reports purpose
+   * @param {Array<String>} emails - A list of user emails relevant to the report
+   * @param {Array<String>} keywords - A list of search terms relevant to the report
+   * @param {Array<String>} spaceNames - A list of space names relevant to the report
+   * @param {Array<String>} spaceIds - A list of space ids relevant to the report
+   * @param {Object} range - Contains the start time and end time defining the search period
+   * @returns {Promise<ResponseEntity>} Http response containing the new report record
+   */
+  @waitForValue('@')
+  createReport(name, description, emails, keywords, spaceNames, spaceIds, range) {
+    if (!emails || !keywords || !spaceNames || !spaceIds || !range) {
+      throw Error('Invalid parameter(s)');
+    }
+
+    const self = this;
+    const encryptionPromise = self._encryptRequestParams(name, description, keywords);
+    const uuidPromise = self._asUUIDs(emails);
+
+    return Promise.all([encryptionPromise, uuidPromise])
+      .then((values) => {
+        const body = reportRequest;
+        body.name = values[0].encryptedName;
+        body.description = values[0].encryptedDescription;
+        body.keywords = values[0].encryptedKeywords;
+        body.encryptionKeyUrl = values[0].encryptionkeyUrl;
+        body.userIds = values[1];
+        body.spaceNames = spaceNames;
+        body.spaceIds = spaceIds;
+        body.range = range;
+
+        return self.request({
+          method: 'POST',
+          service: 'ediscovery',
+          resource: 'reports',
+          body
+        })
+          .then((res) => res)
+          .catch((reason) => {
+            self.logger.error(`Create Report Error ${reason}`);
+            return Promise.reject(reason);
+          });
+      });
+  },
+
+  /**
+   * Retrievse the user UUIDs associated with each supplied email. If an email with no user
+   * is found an error is thrown. If multiple such emails exist only the first encountered
+   * throws an error.
+   *
+   * @param {Array<String>} emails - A list of emails to resolve
+   * @returns {Array<Promise<UUID>>} The UUIDs associated with the emails
+   */
+  _asUUIDs(emails) {
+    if (!emails) {
+      throw Error('Invalid parameter');
+    }
+
+    if (emails.length <= 0) {
+      return Promise.resolve([]);
+    }
+
+    const promises = [];
+    emails.forEach((userEmail) => {
```

Comment:
if `bluebird` is included in the sdk you can use `Promise.map()` to handle some of this broilerplate with pushing promises.

Reply:
Nice, will look into it